### ISSUE TITLE
patch training fails to produce useful pattern

### ISSUE THREAD

**Issue by mkrupczak3** [KI-1-AI-Sec/adversarial-yolo#2]
Currently, patch training is failing to generate an adversarial pattern. It's just optimizing for non-printability score (nps) by generating an image of a single color:

![Screenshot from 2023-10-24 00-28-09](https://github.com/KI-1-AI-Sec/adversarial-yolo/assets/25494111/1f72c6c4-6ac9-4fc0-9b08-1b9ea40c5f88)

@SlicedBacon and I believe this is because the Airbus dataset has negative training images (without objects present) which interfere with the patch training's stochastic gradient descent.

The original paper (and thus, the code repo this one is based on) used the INRIA dataset for person detection. @SlicedBacon noted this dataset does not have any negative training images where a person is not present.

In the current patch training code, any images without objects present will generate a `det_loss` of 0.0, giving the optimizer a false signal that the patch completely defeated the object detector. This seems to be interfering with the patch training process, causing the discontinuities observed in the `det_loss` graph from tensorboard shown above.

For a quick fix, we may remove the negative training images from the dataset when using it for patch training. We should fix it properly so it ignores negative training images in the future though so other researchers can use our code

**Comment by mkrupczak3** [comment]
Removing the negative training images seems to have had little impact. Same result as before.

At our meeting today, discussed how this might be because the patch training dataset is much too large. Theory is that with so many different images, the model is having a hard time learning what works and what doesn't with any single change to the patch.

Going to try a training run on a much smaller subset of data (about 8 images) and if it does any better

**Comment by mkrupczak3** [comment]
Here was the result after 5000 epochs on a set of 8 images:
![5000_epochs](https://github.com/KI-1-AI-Sec/adversarial-yolo/assets/25494111/a42a09e5-48c9-4823-b36a-63e5dba8888e)

Not enough time in the oven. Doing another run over the weekend, this time of 500,000 epochs to see if it works

**Comment by dkyman** [comment]
Hello, I would like to ask why the patch generated by running the code has no gradient. It means that the patch has no changes after multiple rounds of training.I am using the pedestrian data set and the original patch is gray。

**Comment by mkrupczak3** [comment]
It's a known issue, the way we are dealing with the tensors is a bit flawed. There's some significant architectural differences between YOLOv4 and YOLOv8 we haven't been able to work out yet

**Comment by mkrupczak3** [comment]
> Here was the result after 5000 epochs on a set of 8 images: ![5000_epochs](https://private-user-images.githubusercontent.com/25494111/278515068-a42a09e5-48c9-4823-b36a-63e5dba8888e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIzNTk3MjMsIm5iZiI6MTcwMjM1OTQyMywicGF0aCI6Ii8yNTQ5NDExMS8yNzg1MTUwNjgtYTQyYTA5ZTUtNDhjOS00ODIzLWIzNmEtNjNlNWRiYTg4ODhlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjEyVDA1MzcwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ4YWQ3NzE5MjEzNGYzMjI3ZDhkZGMyMDg1MGQ4YzQzNjQ3NzNiOTE2MmFkN2ExODYxZWRkNzdiZjYwYjRiZTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.4UdYm3T1Ty0TWkgpRsaoYGw5xuWraPhJ6rimn-yBRTI)
> 
> Not enough time in the oven. Doing another run over the weekend, this time of 500,000 epochs to see if it works

This didn't do any better; issue is something with the tensors

**Comment by starktao** [comment]
So, does this mean that your project cannot successfully train adversarial examples？

**Comment by mkrupczak3** [comment]
At present, yes. @SlicedBacon has been making some progress but I don't think it's been published here yet

**Comment by starktao** [comment]
That's unfortunate, but thank you.
